content/en/blog/fsm2/index.md

```diff
@@ -1,207 +1,357 @@
+when `data` arrives from `backendChan`.
 
-| Protocol | In/Out | Meaning |
-|----------|--------|---------|
-| `basic_message`| Both | Send or receive messaging protocol |
-| `trust_ping`| Both | we initiate ping protocol |
-| `issue_cred`| Out | we issue a verifiable credential |
-| `present_proof`| Out | we request a proof presentation |
-| `connection`| In | new pairwise is created |
-
-
-### Data
-
-## Issuing Example
-
-The diagram below presents a real-world example of the automatic issuing chat
-bot for verified an email address. Please read carefully the state transition
-arrows. They define triggers and events to send. There is a transition that
-sends an Aries `basic_message` and an `email` in same transition. The email
-message that's built by the machine, includes a random PIN code. As you can see
-the PIN-code can be properly verified in the machine.
-
-{{< figure src="/blog/2023/03/13/finite-state-machine-language-part-i/issue.svg" >}}
-*Automatic Email Credential Chat Bot*
-
-It's been rewarding to notice how well chatting and using verifiable credentials
-fit together. As a end-user you don't face annoying context switches but
-everything is some logical conversation.
-
-## Future Features
-
-- transition triggers are currently SSI-centric which can be changin in future.
-- extremely simple memory model
-    - no persistence model
-- verification/simulation tools
-- simple scriptin language inside the state machine
-- deployment model
-- end-user level tools
+```go
+func (b *Backend) backendReceived(data *fsm.BackendData) {
+	if transition := b.machine.TriggersByBackendData(data); transition != nil {
+		b.send(transition.BuildSendEventsFromBackendData(data))
+		b.machine.Step(transition)
+	}
+}
+```
 
+Both state-machine types (conversation/service level) follow typical transition
+logic:
+1. Do we have a trigger in the current state of the machine?
+1. If the trigger exists, we'll get a transition object for that.
+1. Ask the `transition` to build all send events according to input data.
+1. Send all output events.
+1. If previous steps did succeed, make a state transition step explicit.
+
+### "Channels orchestrate; mutexes serialize"
+
+I'm not a massive fan of the [idiomatic
+Go](https://dave.cheney.net/2020/02/23/the-zen-of-go) or [Go
+proverbs](https://go-proverbs.github.io/). My point is that you shouldn't
+need them. The underlying semantics should be strong enough.
+
+Luckily the world isn't black and white. So, let's use one proverb to state
+something quite obvious.
+
+> "Channels orchestrate; mutexes serialize"
+
+That includes excellent wisdom because it isn't prohibiting you from using the
+mutexes. It clearly states that we need both, but they are used differently.
+That said, the below code block shows the elegance you can achieve with Go
+channels.
+
+```go
+func (c *Conversation) Run(data fsm.MachineData) {
+	c.machine = fsm.NewMachine(data)
```

Comment:
How about explicitly writing with comments here in the code what is the elegance you are after with the statement above?

OR even better, writing example code that does not use channels as a comparison 😅